pages/api/augment-request-gql.ts

```diff
@@ -0,0 +1,408 @@
+  time: ss.number(),
+  identifier: ss.string(),
+  l1RequestTxHash: ss.optional(ss.string()),
+  ooRequestUrl: ss.optional(ss.string()),
+  originatingChainTxHash: ss.optional(ss.string()),
+  originatingChainId: ss.optional(ss.number()),
+  originatingOracleType: ss.optional(ss.string()),
+  optimisticOracleV3Data: ss.optional(
+    ss.object({
+      assertionId: ss.optional(ss.string()),
+      domainId: ss.optional(ss.string()),
+      claim: ss.optional(ss.string()),
+      asserter: ss.optional(ss.string()),
+      callbackRecipient: ss.optional(ss.string()),
+      escalationManager: ss.optional(ss.string()),
+      expirationTime: ss.optional(ss.number()),
+      caller: ss.optional(ss.string()),
+    })
+  ),
+});
+export type ResponseBody = ss.Infer<typeof ResponseBody>;
+
+function extractChildChainId(ancillaryData: string): number | undefined {
+  const childChainIdRegex = /childChainId:(\d+)/;
+  const match = ancillaryData.match(childChainIdRegex);
+  return match ? Number(match[1]) : undefined;
+}
+function constructOoUiLink(
+  txHash: string | undefined,
+  chainId: string | number | undefined,
+  oracleType: string | undefined,
+  eventIndex: string | undefined
+) {
+  if (!txHash || !chainId || !oracleType) return;
+  if (!isSupportedChainId(chainId)) return;
+  const subDomain = Number(chainId) === 5 ? "testnet." : "";
+  return `https://${subDomain}oracle.uma.xyz/request?transactionHash=${txHash}&chainId=${chainId}&oracleType=${castOracleNameForOOUi(
+    oracleType
+  )}&eventIndex=${eventIndex ?? ""}`;
+}
+
+function castOracleNameForOOUi(oracleType: string): string {
+  switch (oracleType) {
+    case "OptimisticOracle":
+      return "Optimistic";
+    case "OptimisticOracleV2":
+      return "OptimisticV2";
+    case "SkinnyOptimisticOracle":
+      return "Skinny";
+    case "OptimisticOracleV3":
+      return "OptimisticV3";
+    default:
+      throw new Error("Unable to cast oracle name for OO UI: " + oracleType);
+  }
+}
+
+async function voteQuery({
+  time,
+  identifier,
+  ancillaryData,
+}: Omit<OracleQueryParams, "chainId">) {
+  const id = makeUniqueKeyForVote(identifier, time, ancillaryData);
+  const query = gql`
+    query voteQuery($id: ID!) {
+      priceRequest(id: $id) {
+        requestTransaction
+      }
+    }
+  `;
+  type GqlRequest = {
+    requestTransaction: string;
+  };
+  type GqlResponse = {
+    priceRequest: GqlRequest | undefined | null;
+  };
+  try {
+    const data: GqlResponse = await request(VoteSubgraphURL, query, { id });
+    assert(data.priceRequest, "request not found");
+    return data?.priceRequest;
+  } catch (error) {
+    if (debug) console.error("vote fetch error:", error);
+    throw error;
+  }
+}
+async function ooSkinnyQuery({
+  time,
+  identifier,
+  chainId,
+}: OracleQueryParams): Promise<ResponseBody> {
+  const subgraph = getSubgraphConfig("Skinny Optimistic Oracle", chainId);
+  const query = gql`
+    query skinnyQuery($proposalTimestamp: Int!) {
```

Comment:
2 price requests can have identical`proposalTimestamp` s

We need to filter in the subgraph by: ancillaryData, time, identifier

Reply:
updated this